### ISSUE TITLE
share plugin not works on iPad

### ISSUE THREAD

**Issue by enricobenedos** [flutter/flutter#47220]
## Steps to Reproduce

1. add share plugin (`share: ^0.6.3+5`) to `pubspec.yaml`
2. main.dart
```dart
import 'package:flutter/material.dart';
import 'package:share/share.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
          child: RaisedButton(
        child: Text('Share Image'),
        onPressed: () async {
          await Share.share(
              'https://www.google.com/url?sa=i&rct=j&q=&esrc=s&source=images&cd=&cad=rja&uact=8&ved=2ahUKEwi-x8ne_LzmAhWC6aQKHZhyAsAQjRx6BAgBEAQ&url=https%3A%2F%2Fmedium.com%2Fflutter-community%2Fmy-thoughts-about-flutter-the-new-cross-platform-framework-bb6c0febf1aa&psig=AOvVaw3RtAhKO3z0Ui4bLOL6dEbm&ust=1576682081381137');
        },
      )),
    );
  }
}
```
3. Try to share the image clicking the button, you will receive this exception:
```
    "<ShareData: 0x6000025870a0>"
)
	Probably at least one of the constraints in the following list is one you don't want.
	Try this:
		(1) look at each constraint and try to figure out which you don't expect;
		(2) find the code that added the unwanted constraint or constraints and fix it.
(
    "<NSLayoutConstraint:0x6000006d9310 LPLinkView:0x7fbbdbd2a840.leading == UILayoutGuide:0x600001c61340'UIViewLayoutMarginsGuide'.leading   (active)>",
    "<NSLayoutConstraint:0x6000006d9b80 H:[LPLinkView:0x7fbbdbd2a840]-(59)-|   (active, names: '|':_UIActivityContentTitleView:0x7fbbdbd06c40 )>",
    "<NSLayoutConstraint:0x6000006ad4f0 H:|-(0)-[_UIActivityContentTitleView:0x7fbbdbd06c40]   (active, names: '|':_UINavigationBarContentView:0x7fbbdbe54140 )>",
    "<NSLayoutConstraint:0x6000006ad540 _UIActivityContentTitleView:0x7fbbdbd06c40.trailing == _UINavigationBarContentView:0x7fbbdbe54140.trailing   (active)>",
    "<NSLayoutConstraint:0x6000006ae8a0 'UIView-Encapsulated-Layout-Width' _UINavigationBarContentView:0x7fbbdbe54140.width == 0   (active)>",
    "<NSLayoutConstraint:0x6000006d9d10 'UIView-leftMargin-guide-constraint' H:|-(16)-[UILayoutGuide:0x600001c61340'UIViewLayoutMarginsGuide'](LTR)   (active, names: '|':_UIActivityContentTitleView:0x7fbbdbd06c40 )>"
)
Will attempt to recover by breaking constraint
<NSLayoutConstraint:0x6000006d9310 LPLinkView:0x7fbbdbd2a840.leading == UILayoutGuide:0x600001c61340'UIViewLayoutMarginsGuide'.leading   (active)>
Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKitCore/UIView.h> may also be helpful.
```

**Target Platform:** iOS
**Target OS version/browser:** Any
**Devices:** iPad 7th (working on iPhone XS)

## Logs

Verbose log is very long, hope it is not a problem attach a dedicated file
[flutter_verbose.log](https://github.com/flutter/flutter/files/3974074/flutter_verbose.log)

```
develops-iMac:broken_share_plugin develop$ flutter analyze
Analyzing broken_share_plugin...                                        
No issues found! (ran in 1.4s)
develops-iMac:broken_share_plugin develop$ 
```

```
[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.15.2 19C57, locale en-IT)
    • Flutter version 1.12.13+hotfix.5 at /Users/develop/SDKs/flutter
    • Framework revision 27321ebbad (7 days ago), 2019-12-10 18:15:01 -0800
    • Engine revision 2994f7e1e6
    • Dart version 2.7.0

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/develop/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3, Build version 11C29
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 38.1.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[✓] VS Code (version 1.41.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.7.1

• No issues found!
```

**Comment by muhammadarslan** [comment]
What is the status? is someone working on this?

**Comment by amenk** [comment]
Also reproducible on iPad Pro 12.0 inch 3rd generation in the simulator

**Comment by mdoelker** [comment]
Specifying the optional `sharePositionOrigin` fixes it. I guess there should be some kind of default in place in the plugin.

**Comment by mdoelker** [comment]
https://github.com/flutter/flutter/issues/44962

**Comment by iapicca** [comment]
Hi @enricobenedos 
I see there's an open [issue](https://github.com/flutter/flutter/issues/19152) addressing the case you described.
Please follow up on that issue,
I'm closing the current one as duplicate.
If you disagree please write in the comments
and I will reopen it.
Thank you

**Comment by enricobenedos** [comment]
Hey @iapicca, I check the issue you link me. It seems different, in my case the app does not crash. Simply on the iPad the "share menù" is not showed.
In my case the iOS version is the `13.3.1`. The @mdoelker workaround works great but by me this is a package regression because in the previous versions all works well.

**Comment by abegehr** [comment]
That `sharePositionOrigin` parameter is needed for iPads, should be mentioned in the docs of the share plugin: https://github.com/flutter/plugins/tree/master/packages/share

**Comment by enricobenedos** [comment]
Hi @abegehr, I'm not able to find the documentation you are telling me. In the README and the CHANGELOG it is not mentioned. Which documentation are you meaning?

**Comment by dahabit** [comment]
My issue was my share button changeable and inside Grid view and most solutions displays the share box in a fixed place, I used https://pub.dev/packages/rect_getter to get my Widget position on the screen then show the share box based on that and now I can display the share button exactly beside my item on iPad 
steps 

1- install the package [https://pub.dev/packages/rect_getter ](https://pub.dev/packages/rect_getter )

2- add `final globalKey = RectGetter.createGlobalKey();` at the begining of your widget

3- on your share button click event add 
`onPressed: () {
                              **var rect = RectGetter.getRectFromKey(globalKey);**
                              Share.share('Title',subject: 'shareUrl',
                                **sharePositionOrigin: Rect.fromLTWH(
                                    rect.left + 40, rect.top + 20, 2, 2),**
                              );
                            },

I hope this temporary fix helps anyone

**Comment by angel1st** [comment]
Definitely, Flutter team has to update the doc. Their [example](https://github.com/flutter/plugins/blob/master/packages/share/example/lib/main.dart) in fact does not show the share popup properly on iPads. 

The way I handle it was w/o any additional libraries. 
Here it is:
 
```
final Size size = MediaQuery.of(context).size;
......
onPressed: () => Share.share(
     message,
     sharePositionOrigin: Rect.fromLTWH(0, 0, size.width, size.height / 2),
),
```

**Comment by enricobenedos** [comment]
@angel1st I agree with you. The strange thing is that on previous versions of this package, it works "perfectly" also on iPad

**Comment by angel1st** [comment]
Then I guess this is a regression bug and has to be fixed. It is very basic issue, easy to catch, hence no excuse for being around for more that a quarter!

**Comment by abegehr** [comment]
> Hi @abegehr, I'm not able to find the documentation you are telling me. In the README and the CHANGELOG it is not mentioned. Which documentation are you meaning?

hey @enricobenedos, yes, I mean the README.md: https://github.com/flutter/plugins/tree/master/packages/share#example

**Comment by AirborneEagle** [comment]
My Production app just had its latest update rejected because of this!
the person reviewing the app just happened to be using an iPad, and rejected it.
Ridiculous.
@angel1st  thank you for your simple workaround. 👍

**Comment by egphp** [comment]
> final Size size = MediaQuery.of(context).size;

thanks . that's help me

**Comment by computerzman** [comment]
> Definitely, Flutter team has to update the doc. Their [example](https://github.com/flutter/plugins/blob/master/packages/share/example/lib/main.dart) in fact does not show the share popup properly on iPads.
> 
> The way I handle it was w/o any additional libraries.
> Here it is:
> 
> ```
> final Size size = MediaQuery.of(context).size;
> ......
> onPressed: () => Share.share(
>      message,
>      sharePositionOrigin: Rect.fromLTWH(0, 0, size.width, size.height / 2),
> ),
> ```

Thanks, this worked for me ...

**Comment by gapspt** [comment]
I have a fix for this and I am about to make a pull request.

**Comment by gapspt** [comment]
> I have a fix for this and I am about to make a pull request.
I am actually waiting for another pull request to be merged first in order to proceed to this one.

**Comment by alvinkonda** [comment]
@AirborneEagle They always use iPads, I got rejected too. Tested with iPhone, all fine. Definitely the docs should be updated

**Comment by eseidelGoogle** [comment]
@gapspt are you still waiting on a review?  Thanks.

**Comment by devxpy** [comment]
Here's a solution that shows the share popup at the correct location.

```dart
final shareButtonKey = GlobalKey();

Widget build(BuildContext context) {
	...
	IconButton(
	  key: shareButtonKey,
	  icon: Icon(Icons.share),
	  onPressed: shareFile,
	)
}

void shareFile() {
  Share.shareFiles(
	...
    sharePositionOrigin: shareButtonRect(),
  );
}

Rect shareButtonRect() {
  RenderBox renderBox = shareButtonKey.currentContext?.findRenderObject();
  if (renderBox == null) return null;

  Size size = renderBox.size;
  Offset position = renderBox.localToGlobal(Offset.zero);

  return Rect.fromCenter(
    center: position + Offset(size.width / 2, size.height / 2),
    width: size.width,
    height: size.height,
  );
}
```

**Comment by TahaTesser** [comment]
Can confirm, this is fixed, tested on my new iPad on the latest `share: ^0.6.5+4` 
### Preview 
![IMG_0053](https://user-images.githubusercontent.com/48603081/97433052-e5456b80-1942-11eb-8668-0ed62df9193a.jpg)


### Code Sample
[Official Sample](https://pub.dev/packages/share/example)

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 1.22.2, on Mac OS X 10.15.7 19H2, locale en-GB)
    • Flutter version 1.22.2 at /Users/tahatesser/Code/flutter_stable
    • Framework revision 84f3d28555 (13 days ago), 2020-10-15 16:26:19 -0700
    • Engine revision b8752bbfff
    • Dart version 2.10.2

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    • CocoaPods version 1.10.0.rc.1

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.1

[✓] Connected device (1 available)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E • ios • iOS 14.1

! Doctor found issues in 1 category.
```

</details>

**Comment by enricobenedos** [comment]
It work better even if it is not perfectly aligned

<img width="136" alt="image" src="https://user-images.githubusercontent.com/12233746/97435942-786da980-1921-11eb-9ee1-01d019f827e2.png">

```dart
appBar: AppBar(
        actions: [
              IconButton(
      icon: Platform.isIOS
          ? const Icon(
              CupertinoIcons.share,
            )
          : const Icon(
              Icons.share,
            ),
      onPressed: () async {
        //Simply share the link as string
        print('Sharing image -->' + imgUrl);
        if (imgUrl != null && imgUrl.isNotEmpty) {
          final RenderBox box = context.findRenderObject();

          await Share.share(
            imgUrl,
            sharePositionOrigin: box.localToGlobal(Offset.zero) & box.size,
          );
        }
      },
    )],
      ),
```

**Comment by TahaTesser** [comment]
@enricobenedos
Please file a new issue for improvements

**Comment by jameslester2004** [comment]
> Here's a solution that shows the share popup at the correct location.
> 
> ```dart
> final shareButtonKey = GlobalKey();
> 
> Widget build(BuildContext context) {
> 	...
> 	IconButton(
> 	  key: shareButtonKey,
> 	  icon: Icon(Icons.share),
> 	  onPressed: shareFile,
> 	)
> }
> 
> void shareFile() {
>   Share.shareFiles(
> 	...
>     sharePositionOrigin: shareButtonRect(),
>   );
> }
> 
> Rect shareButtonRect() {
>   RenderBox renderBox = shareButtonKey.currentContext?.findRenderObject();
>   if (renderBox == null) return null;
> 
>   Size size = renderBox.size;
>   Offset position = renderBox.localToGlobal(Offset.zero);
> 
>   return Rect.fromCenter(
>     center: position + Offset(size.width / 2, size.height / 2),
>     width: size.width,
>     height: size.height,
>   );
> }
> ```

Thanks. This was the only solution that actually displayed the box in the right place. @angel1st solution was progress, but had the box showing up in the middle of the screen. Just calling context.findRenderObject(); as the docs state resulted in the box showing off the screen in the top left, so not sure if this has been fixed properly. By using the global key, it seems to locate the trigger button and display the box below it. 

` _onShare(BuildContext context, String link) async {

 RenderBox renderBox = shareButtonKey.currentContext?.findRenderObject();

    Share.share('Check out this competition on Golfify $link',
        sharePositionOrigin:
            renderBox.localToGlobal(Offset.zero) & renderBox.size,
        subject: 'New Golfify Competition');
  }
